src/stream_base.cc

```diff
@@ -0,0 +1,504 @@
+void StreamBase::JSMethod(const FunctionCallbackInfo<Value>& args) {
+  StreamBase* wrap = Unwrap<Base>(args.Holder());
+
+  if (!wrap->IsAlive())
+    return args.GetReturnValue().Set(UV_EINVAL);
+
+  args.GetReturnValue().Set((wrap->*Method)(args));
+}
+
+
+int StreamBase::ReadStart(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  return ReadStart();
+}
+
+
+int StreamBase::ReadStop(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  return ReadStop();
+}
+
+
+int StreamBase::Shutdown(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsObject());
+  Local<Object> req_wrap_obj = args[0].As<Object>();
+
+  ShutdownWrap* req_wrap = new ShutdownWrap(env,
+                                            req_wrap_obj,
+                                            this,
+                                            AfterShutdown);
+
+  int err = DoShutdown(req_wrap);
+  req_wrap->Dispatched();
+  if (err)
+    delete req_wrap;
+  return err;
+}
+
+
+void StreamBase::AfterShutdown(ShutdownWrap* req_wrap, int status) {
+  StreamBase* wrap = req_wrap->wrap();
+  Environment* env = req_wrap->env();
+
+  // The wrap and request objects should still be there.
+  CHECK_EQ(req_wrap->persistent().IsEmpty(), false);
+  CHECK_EQ(wrap->GetObject().IsEmpty(), false);
+
+  HandleScope handle_scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  Local<Object> req_wrap_obj = req_wrap->object();
+  Local<Value> argv[3] = {
+    Integer::New(env->isolate(), status),
+    wrap->GetObject(),
+    req_wrap_obj
+  };
+
+  if (req_wrap->object()->Has(env->oncomplete_string()))
+    req_wrap->MakeCallback(env->oncomplete_string(), ARRAY_SIZE(argv), argv);
+
+  delete req_wrap;
+}
+
+
+int StreamBase::Writev(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args[0]->IsObject());
+  CHECK(args[1]->IsArray());
+
+  Local<Object> req_wrap_obj = args[0].As<Object>();
+  Local<Array> chunks = args[1].As<Array>();
+
+  size_t count = chunks->Length() >> 1;
+
+  uv_buf_t bufs_[16];
+  uv_buf_t* bufs = bufs_;
+
+  // Determine storage size first
+  size_t storage_size = 0;
+  for (size_t i = 0; i < count; i++) {
+    Handle<Value> chunk = chunks->Get(i * 2);
+
+    if (Buffer::HasInstance(chunk))
+      continue;
+      // Buffer chunk, no additional storage required
+
+    // String chunk
+    Handle<String> string = chunk->ToString(env->isolate());
+    enum encoding encoding = ParseEncoding(env->isolate(),
+                                           chunks->Get(i * 2 + 1));
+    size_t chunk_size;
+    if (encoding == UTF8 && string->Length() > 65535)
+      chunk_size = StringBytes::Size(env->isolate(), string, encoding);
+    else
+      chunk_size = StringBytes::StorageSize(env->isolate(), string, encoding);
+
+    storage_size += chunk_size + 15;
```

Comment:
This is a question for a separate issue, I was only migrating old code base here too. :) I'm afraid I don't even remember where did this came from, but it might be not needed anymore...